lib/Sema/TypeCheckProtocolInference.cpp

```diff
@@ -1204,27 +1200,54 @@ void AssociatedTypeInference::findSolutionsRec(
       return;
     }
 
-    /// Check the current set of type witnesses.
-    bool invalid = checkCurrentTypeWitnesses(valueWitnesses);
+    ++NumSolutionStates;
+
+    // Fold the dependent member types within this type.
+    for (auto assocType : proto->getAssociatedTypeMembers()) {
+      if (conformance->hasTypeWitness(assocType))
+        continue;
 
-    // Determine whether there is already a solution with the same
-    // bindings.
-    for (const auto &solution : solutions) {
-      bool sameSolution = true;
+      // If the type binding does not have a type parameter, there's nothing
+      // to do.
+      auto known = typeWitnesses.begin(assocType);
+      assert(known != typeWitnesses.end());
+      if (!known->first->hasTypeParameter() &&
+          !known->first->hasDependentMember())
+        continue;
+
+      Type replaced = substCurrentTypeWitnesses(known->first);
+      if (replaced.isNull())
+        return;
+
+      known->first = replaced;
+    }
+
+    // Check whether our current solution matches the given solution.
+    auto matchesSolution =
+        [&](const InferredTypeWitnessesSolution &solution) {
       for (const auto &existingTypeWitness : solution.TypeWitnesses) {
         auto typeWitness = typeWitnesses.begin(existingTypeWitness.first);
-        if (!typeWitness->first->isEqual(existingTypeWitness.second.first)) {
-          sameSolution = false;
-          break;
-        }
+        if (!typeWitness->first->isEqual(existingTypeWitness.second.first))
+          return false;
       }
 
-      // We found the same solution. There is no point in recording
-      // a new one.
-      if (sameSolution)
-        return;
+      return true;
+    };
+
+    // If we've seen this solution already, bail out; there's no point in
+    // checking further.
+    if (std::find_if(solutions.begin(), solutions.end(), matchesSolution)
```

Comment:
FWIW, I think these could probably be `llvm::all_of(solutions, matchesSolution)`.

Reply:
You mean `llvm::any_of`, I presume. I didn't know about that one, thanks!